File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool solution(vector<string> phone_book) {
    bool answer = true;
    std::sort(phone_book.begin(), phone_book.end());
    // 접두어 있으면 false
    for (int i = 0; i < phone_book.size()-1; i++) {
        int len = std::min(phone_book[i].size(), phone_book[i+1].size());
        for (int j = 0; j < len; j++) {
            if (phone_book[i][j] != phone_book[i+1][j]) {
                break;
            }
            if (j == len-1){
                answer = false;
            }
        }
        if (!answer) {
            break;
        }
    }
    return answer;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        // HashMap<String, String> hm = new HashMap<>();
        // for (int i = 0; i < phone_book.length; i++) {
        //     for (int j = 0; j < phone_book[i].length(); j++) {
        //         if (hm.containsKey(phone_book[i].substring(0, j+1))) {
        //             return false;
        //         }
        //         hm.put(phone_book[i].substring(0, j+1), phone_book[i]);
        //     }   
        // }
        Arrays.sort(phone_book);
        for (int i = 0; i < phone_book.length-1; i++) {
            if (phone_book[i+1].startsWith(phone_book[i])) {
                return false;
            }
        }

        return answer;
    }
}